add transpiler to doc (#275)
Qiskit/qiskit: docs/terra/transpiler/about/pass_managers.rst

--------------------
Control Flow Plugins
--------------------

By default, there are two control flow plugins included in the default pass
manager: ``do_while`` and ``conditional`` (see **Fixed Point** and
**Conditional** use cases). You might want to add more control flow plugins. For
example, a for-loop can be implemented in the following way:

.. code:: python

  class DoXTimesController(FlowController):
    def __init__(self, passes, do_x_times, **_):
      self.do_x_times = do_x_times()
      super().__init__(passes)

    def __iter__(self):
      for _ in range(self.do_x_times):
        for pass_ in self.working_list:
          yield pass_

The plugin is added to the pass manager in this way:

.. code:: python

  self.passmanager.add_flow_controller('do_x_times', DoXTimesController)

This allows to use the parameter ``do_x_times``, which needs to be a callable.
In this case, this is used to parametrized the plugin, so it will for-loop 3
times.

.. code:: python

  self.passmanager.append([Pass()], do_x_times=lambda x : 3)

---------
Use Cases
---------

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Simple Chain with Dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The ``CxCancellation`` requires and preserves ``ToffoliDecompose``. Same for

Reviewer: ```suggestion
The ``CXCancellation`` requires and preserves ``Decompose``. Same for
```

> ``RotationMerge``. The pass ``Mapper`` requires extra information for running

Reviewer: ```suggestion
``Optimize1qGates``. The pass ``BasicSwap`` requires extra information for running
```

(the ``coupling_map``, in this case).

.. code:: python

  pm = PassManager()

>   pm.append(CxCancellation()) # requires:  ToffoliDecompose

Reviewer: ```suggestion
  pm.append(CXCancellation()) # requires:  Decompose
```

>                               # preserves: ToffoliDecompose

Reviewer: ```suggestion
                              # preserves: Decompose
```

>   pm.append(RotationMerge())  # requires:  ToffoliDecompose

Reviewer: ```suggestion
  pm.append(Optimize1qGates())  # requires:  Decompose
```

>                               # preserves: ToffoliDecompose

Reviewer: ```suggestion
                              # preserves: Decompose
```

  pm.append(Mapper(coupling_map=coupling_map)) # requires:  []
                                               # preserves: []

>   pm.append(CxCancellation())

Reviewer: ```suggestion
  pm.append(CXCancellation())
```

Given the above, the pass manager executes the following sequence of passes:

> #. ``ToffoliDecompose``, because it is required by ``CxCancellation``.

Reviewer: these are to make the pass names consistent with terra.
```suggestion
#. ``ToffoliDecompose``, because it is required by ``CXCancellation``.
```

> #. ``ToffoliDecompose``, because it is required by ``CxCancellation``.

Reviewer: ```suggestion
#. ``Decompose``, because it is required by ``CXCancellation``.
```

#. ``CxCancellation``

> #. ``RotationMerge``, because, even when ``RotationMerge`` also requires

Reviewer: ```suggestion
#. ``Optimize1qGates``, because even though ``Optimize1qGates`` also requires
```

>    ``ToffoliDecompose``, the ``CxCancellation`` preserved it, so no need to run

Reviewer: ```suggestion
   ``Decompose``, the ``CXCancellation`` preserved it, so no need to run
```

   it again.

> #. ``Mapper``

Reviewer: ```suggestion
```

> #. ``ToffoliDecompose``, because ``Mapper`` did not preserved

Reviewer: ```suggestion
#. ``ToffoliDecompose``, because ``Mapper`` did not preserve
```

>    ``ToffoliDecompose`` and is require by ``CxCancellation``

Reviewer: ```suggestion
   ``Decompose`` and it is required by ``CxCancellation``
```

> #. ``CxCancellation``

Reviewer: ```suggestion
```

unrolling using some basis gates is totally different than unrolling to
different gates. And a PassManager might use both.

.. code:: python

  pm.append(Unroller(basis_gates=['id','u1','u2','u3','cx']))
  pm.append(...)
  pm.append(Unroller(basis_gates=['U','CX']))

where (from ``qelib1.inc``):

.. code:: python

  gate id q { U(0,0,0) q; }
  gate u1(lambda) q { U(0,0,lambda) q; }
  gate u2(phi,lambda) q { U(pi/2,phi,lambda) q; }
  gate u3(theta,phi,lambda) q { U(theta,phi,lambda) q; }
  gate cx c,t { CX c,t; }

For this reason, the identity of a pass is given by its name and parameters.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^
While Loop up to Fixed Point
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

There are cases when one or more passes have to be run repeatedly, until a
condition is fulfilled.

.. code:: python

  pm = PassManager()
  pm.append([CxCancellation(), RotationMerge(), CalculateDepth()],
    do_while=lambda property_set: not property_set['fixed_point']['depth'])

The control argument ``do_while`` will run these passes until the callable
returns ``False``. The callable always takes in one argument, the pass manager's

> property set. In this example, ``CalculateDepth`` is an analysis pass that

Reviewer: ```suggestion
property set. In this example, ``Depth`` is an analysis pass that
```

updates the property ``depth`` in the property set.

^^^^^^^^^^^
Conditional
^^^^^^^^^^^

The pass manager developer can avoid one or more passes by making them
conditional (on a property in the property set):

.. code:: python

>   pm.append(BasicLayout(coupling_map))

Reviewer: ```suggestion
  pm.append(TrivialLayout(coupling_map))
```

  pm.append(CheckMap(coupling_map))
  pm.append(BasicSwap(coupling_map),
    condition=lambda property_set: not property_set['is_swap_mapped'])

The ``CheckMap`` is an analysis pass that updates the property

> ``is_swap_mapped``. If ``LayoutMapper`` could map the circuit to the coupling

Reviewer: ```suggestion
``is_swap_mapped``. If ``TrivialLayout`` could map the circuit to the coupling
```

> map, the ``SwapMapper`` is unnecessary.

Reviewer: ```suggestion
map, the ``BasicSwap`` is unnecessary.
```
